paasta_tools/docker_wrapper.py

```diff
@@ -71,31 +76,105 @@ def generate_hostname(fqdn, mesos_task_id):
 
 
-def add_hostname(args, hostname):
-    # Add --hostname argument immediately after 'run' command if it exists
+def add_argument(args, argument):
+    # Add an argument immediately after 'run' command if it exists
     args = list(args)
-
     try:
         run_index = args.index('run')
     except ValueError:
         pass
     else:
-        args.insert(run_index + 1, '--hostname=' + hostname)
-
+        args.insert(run_index + 1, argument)
     return args
 
 
+def get_cpumap():
+    # Return a dict containing the core numbers per physical CPU
+    core = 0
+    cpumap = {}
+    try:
+        with open('/proc/cpuinfo', 'r') as f:
+            for line in f:
+                m = re.match('physical\sid.*(\d)', line)
+                if m:
+                    cpuid = int(m.group(1))
+                    if cpuid not in cpumap:
+                        cpumap[cpuid] = []
+                    cpumap[cpuid].append(core)
+                    core += 1
+    except IOError:
+        pass
+    return cpumap
+
+
+def is_numa_enabled():
+    return os.path.exists('/proc/1/numa_maps')
+
+
+def arg_collision(new_args, current_args):
+    # Returns True if one of the new arguments is already in the
+    # current argument list.
+    cur_arg_keys = []
+    for c in current_args:
+        cur_arg_keys.append(c.split('=')[0])
+    return bool(set(new_args).intersection(set(cur_arg_keys)))
+
+
+def validate_int(i):
+    try:
+        return int(i)
+    except (ValueError, TypeError):
+        return None
+
+
+def validate_float(f):
+    # Ensure we return a float. If input is invalid we return 0.0
+    try:
+        return float(f)
+    except (ValueError, TypeError):
+        return 0.0
+
+
+def generate_cpuset_args(env_args, pinned_numa_node):
+    required_cores = validate_float(env_args.get('MARATHON_APP_RESOURCE_CPUS'))
+    cpumap = get_cpumap()
+    result = []
+
+    # No need to set up docker cpuset options if one of the above fails:
+    if (
+        # NUMA is supported by the system
+        is_numa_enabled() and
+        # Machine has multiple CPUs
+        len(cpumap) > 1 and
+        # Requested numa node exists
+        pinned_numa_node in cpumap and
+        # The numa node has more cores than the container requires
+        len(cpumap[pinned_numa_node]) >= required_cores
+    ):
+        result.append('--cpuset-cpus=' + ','.join(str(c) for c in cpumap[pinned_numa_node]))
+        result.append('--cpuset-mems=' + str(pinned_numa_node))
+
+    return result
+
+
 def main(argv=None):
     argv = argv if argv is not None else sys.argv
 
+    # Get Docker env variables
     env_args = parse_env_args(argv)
-    fqdn = socket.getfqdn()
 
     # Marathon sets MESOS_TASK_ID whereas Chronos sets mesos_task_id
     mesos_task_id = env_args.get('MESOS_TASK_ID') or env_args.get('mesos_task_id')
 
+    # Check if we require this container to stick to a given NUMA node
+    pinned_numa_node = validate_int(env_args.get('PIN_TO_NUMA_NODE'))
```

Comment:
The validate_int stuff is kinda puppety and not very pythony.

Please just return None and key off that and only turn to `int()` at the last second.

I want this whole block to be only 2 lines:

```
if env_get.get('Pin..):
    argv = append_cpuset_args(argv, env_get.get('Pin..))
```

For 2 reasons:
1. The more code out of main means it is under test
2. The the non-pinning code path "looks" even smaller and smaller" if you are stepping through the code with your eyes